File: wasm/ggml-shielded/shielded-probe.c
/*
 * shielded-probe -- one real masked GEMM through the C stack, asserted four ways.
 *
 * The C counterpart of metal/guest/shielded-probe.mjs, and the thing that proves
 * shielded-field/-wire/-tee actually talk to a live worker rather than merely
 * compiling. It asserts, against the bytes that really crossed the socket:
 *   exact         the unmasked product equals a local int64 reference
 *   verified      Freivalds accepts the honest result
 *   lie_rejected  Freivalds rejects a single-element corruption
 *   denylist      the worker refuses a denylisted op ON THE WIRE
 *
 * The FIXTURE is a seeded LCG so a failure reproduces from the log. The Freivalds
 * secret is NOT part of the fixture -- it comes from the OS CSPRNG inside
 * shielded-tee, because "lie_rejected" only means something if the worker could
 * not have predicted s.
 */
#include "shielded-field.h"
#include "shielded-tee.h"
#include "shielded-wire.h"

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

static uint32_t seed = 0x2f6e2b1;
static double rnd(void) { seed = (uint32_t)(seed * 1103515245u + 12345u) & 0x7fffffff; return (double)seed / (double)0x7fffffff; }

static uint16_t float_to_half(float v) {
    uint32_t x; memcpy(&x, &v, 4);
    uint32_t sign = (x >> 16) & 0x8000;
    int exp = (int)((x >> 23) & 0xff) - 127 + 15;
    uint32_t man = (x >> 13) & 0x3ff;
    if (exp <= 0) {
        int shift = 1 - exp;
        if (shift > 24) return (uint16_t)sign;
        man = ((x & 0x7fffff) | 0x800000) >> (13 + shift);
        return (uint16_t)(sign | man);
    }
    if (exp >= 31) return (uint16_t)(sign | 0x7c00);
    return (uint16_t)(sign | ((uint32_t)exp << 10) | man);
}

int main(int argc, char **argv) {
    const char *host = "127.0.0.1";
    int port = 9500, K = 512, N = 256, m = 1;
    for (int i = 1; i < argc - 1; i++) {
        if (!strcmp(argv[i], "--host")) host = argv[++i];
        else if (!strcmp(argv[i], "--port")) port = atoi(argv[++i]);
        else if (!strcmp(argv[i], "--k")) K = atoi(argv[++i]);
        else if (!strcmp(argv[i], "--n")) N = atoi(argv[++i]);
    }

    int8_t   *wq = malloc((size_t)K * N);
    uint16_t *wd = malloc((size_t)(K / SH_QK) * N * 2);
    int64_t  *x  = malloc((size_t)m * K * sizeof(int64_t));
    int64_t  *y  = malloc((size_t)m * N * sizeof(int64_t));
    int64_t  *want = malloc((size_t)m * N * sizeof(int64_t));
    int8_t   *wf = malloc((size_t)K * N);
    if (!wq || !wd || !x || !y || !want || !wf) { fprintf(stderr, "oom\n"); return 2; }

    for (int i = 0; i < (K / SH_QK) * N; i++) wd[i] = float_to_half((float)(0.001 + rnd() * 0.0025));
    for (int i = 0; i < K * N; i++) wq[i] = (int8_t)lrint((rnd() * 2 - 1) * 127);
    for (int i = 0; i < m * K; i++) x[i] = (int64_t)lrint((rnd() * 2 - 1) * 900);

    /* Independent reference: recompute the encoding and the product in int64. */
    for (int k = 0; k < K; k++)
        for (int j = 0; j < N; j++)
            wf[(size_t)k * N + j] = (int8_t)sh_encode_weight_fixed(wd[(size_t)(k / SH_QK) * N + j],
                                                                   wq[(size_t)k * N + j]);
    int64_t peak = 0;
    for (int i = 0; i < m; i++)
        for (int j = 0; j < N; j++) {
            int64_t acc = 0;
            for (int k = 0; k < K; k++) acc += x[(size_t)i * K + k] * wf[(size_t)k * N + j];
            want[(size_t)i * N + j] = acc;
            int64_t a = acc < 0 ? -acc : acc;
            if (a > peak) peak = a;
        }

    int err = SH_OK;
    sh_link *l = sh_link_open(host, port, true, &err);
    if (!l) { fprintf(stderr, "open failed\n"); return 2; }
    int node = sh_link_add_weight(l, "probe", wq, wd, K, N, m, -1);
    if (node < 0) { fprintf(stderr, "add_weight: %s\n", sh_link_last_error(l)); return 2; }
    if ((err = sh_link_start(l)) != SH_OK) { fprintf(stderr, "start: %s\n", sh_link_last_error(l)); return 2; }

    int64_t *outs[1] = { y };
    int nodes[1] = { node };
    if ((err = sh_link_gemm(l, nodes, 1, x, m, outs)) != SH_OK) {
        fprintf(stderr, "gemm: %s\n", sh_link_last_error(l)); return 2;
    }

    int exact = 1;
    for (int i = 0; i < m * N; i++) if (y[i] != sh_balanced(want[i])) { exact = 0; break; }

    /* Both directions, against the same checker the online path uses. Asserting
     * only that it ACCEPTS the honest product would pass just as happily if the
     * check were `return true`. */
    int verified = sh_link_verify(l, node, x, y, m) ? 1 : 0;
    int64_t save = y[N / 2];
    y[N / 2] = save + 1;                /* a single-element lie */
    int lie_rejected = sh_link_verify(l, node, x, y, m) ? 0 : 1;
    y[N / 2] = save;

    uint64_t ex = 0, macs = 0, vf = 0;
    sh_link_stats(l, &ex, &macs, &vf);
    sh_link_close(l);

    /* The denylist, on the wire, on its own connection (install is once-only). */
    int denylist = 0;
    sh_pipe *p = sh_pipe_open(host, port, &err);
    if (p) {
        uint8_t pay[64]; sh_reply rep;
        size_t n = sh_pack_hello(pay, 1);
        if (sh_pipe_call(p, SH_CMD_HELLO, pay, n, &rep) == SH_OK) {
            sh_reply_free(&rep);
            n = sh_pack_alloc(pay, 4096, "activations");
            if (sh_pipe_call(p, SH_CMD_ALLOC_BUFFER, pay, n, &rep) == SH_OK) sh_reply_free(&rep);
            const char *bad = "{\"nodes\":[{\"op\":\"SOFT_MAX\"}],"
                              "\"outputs\":[{\"bid\":1,\"offset\":0,\"nbytes\":16}]}";
            int rc = sh_pipe_call(p, SH_CMD_GRAPH_INSTALL, bad, strlen(bad), &rep);
            denylist = (rc == SH_ERR_VIOLATION);
            if (denylist) fprintf(stderr, "worker refused as expected: %s\n", sh_pipe_last_error(p));
            if (rc == SH_OK) sh_reply_free(&rep);
        }
        sh_pipe_close(p);
    }

    printf("{\"exact\":%s,\"verified\":%s,\"lie_rejected\":%s,\"denylist_refused\":%s,"
           "\"peak_abs_y\":%lld,\"field_headroom\":%.2f,\"K\":%d,\"N\":%d,"
           "\"exchanges\":%llu,\"macs\":%llu,\"verify_fail\":%llu}\n",
           exact ? "true" : "false", verified ? "true" : "false",
           lie_rejected ? "true" : "false", denylist ? "true" : "false",
           (long long)peak, (double)SH_HALF_M / (double)(peak ? peak : 1), K, N,
           (unsigned long long)ex, (unsigned long long)macs, (unsigned long long)vf);
    return (exact && verified && lie_rejected && denylist && vf == 0) ? 0 : 1;
}

File: metal/guest/shielded-probe.mjs
#!/usr/bin/env node
// shielded-probe.mjs — the metal box's proof that its GPU is usable WITHOUT
// being trusted. Runs inside the CVM; talks to shielded/worker.py on the host.
//
//   node /opt/metal/shielded-probe.mjs --host 10.0.2.2 --port 9500
//
// It performs one real masked field GEMM end to end and asserts four things:
//
//   exact        the unmasked product equals the plaintext product, bit for bit.
//                Slalom recovery is exact in Z_M, so anything less is a bug.
//   verified     preprocessed Freivalds accepts the honest result, and REJECTS a
//                single-element lie -- the hardest case to catch, and the one a
//                lazy verifier passes.
//   no-plaintext no word of the secret activation appears anywhere in the bytes
//                that crossed the boundary. This is checked against the actual
//                transcript, not argued.
//   refused      the worker rejects a graph containing an op from the denylist,
//                so the admission rules are live on the wire and not just in
//                protocol.py's unit test.
//
// Exits non-zero if any of them fails, so it can gate a box advertising the
// shielded flavor.

import {
  CMD, Freivalds, HALF_M, M_MOD, MaskBank, QK, ShieldedLink, crt,
  encodeWeightFixed, floatToHalf, packAlloc, packRecompute, packRegion, packSet,
  refill, toResidues,
} from './shielded.mjs';

const arg = (n, d) => { const i = process.argv.indexOf('--' + n); return i > 0 ? process.argv[i + 1] : d; };
const HOST = arg('host', '10.0.2.2');
const PORT = Number(arg('port', 9500));
const K = Number(arg('k', 512));
const N = Number(arg('n', 256));

function frame(cmd, payload = Buffer.alloc(0)) {
  const h = Buffer.alloc(9); h[0] = cmd; h.writeBigUInt64LE(BigInt(payload.length), 1);
  return Buffer.concat([h, payload]);
}

// Deterministic PRNG so a failure is reproducible from the log alone.
let seed = 0x2f6e2b1;
const rnd = () => { seed = (seed * 1103515245 + 12345) & 0x7fffffff; return seed / 0x7fffffff; };

async function main() {
  const out = { host: `${HOST}:${PORT}`, K, N };

  // ---- public weights: q8_0 quants plus one fp16 scale per 32 rows ----------
  const wq = new Int8Array(K * N);
  const wdHalf = new Uint16Array((K / QK) * N);
  for (let i = 0; i < wq.length; i++) wq[i] = Math.round((rnd() * 2 - 1) * 127);
  for (let i = 0; i < wdHalf.length; i++) wdHalf[i] = floatToHalf(0.001 + rnd() * 0.0025);
  const wFixed = new Float64Array(K * N);
  let peakW = 0;
  for (let k = 0; k < K; k++) {
    for (let j = 0; j < N; j++) {
      const v = encodeWeightFixed(wdHalf[Math.floor(k / QK) * N + j], wq[k * N + j]);
      wFixed[k * N + j] = v;
      peakW = Math.max(peakW, Math.abs(v));
    }
  }
  out.peak_w_fixed = peakW;
  if (peakW > 119) throw new Error(`|w_fixed| ${peakW} exceeds the kernel's byte lane`);

  // ---- the secret ----------------------------------------------------------
  const x = new Float64Array(K);
  for (let k = 0; k < K; k++) x[k] = Math.round((rnd() * 2 - 1) * 900);

  // Wait for the worker rather than racing it. The guest reaches userspace in a
  // few seconds; the worker spends ~10 s importing torch before it binds. On a
  // cold boot the probe would therefore hit ECONNREFUSED and report the card
  // missing when it is merely still starting -- a false negative that would take
  // a healthy box out of the shielded tier. Bounded, so a genuinely absent worker
  // still fails rather than hanging boot.
  const DEADLINE_MS = Number(arg('wait-ms', 90000));
  const link = new ShieldedLink(HOST, PORT);
  const t_wait0 = Date.now();
  for (let attempt = 0; ; attempt++) {
    try { await link.connect(); break; } catch (e) {
      if (Date.now() - t_wait0 > DEADLINE_MS) throw e;
      await new Promise((r) => setTimeout(r, Math.min(1000 * (attempt + 1), 5000)));
    }
  }
  out.waited_ms = Date.now() - t_wait0;
  const hello = JSON.parse((await link.call(CMD.HELLO, Buffer.from([1, 0, 0, 0]))).toString());
  const GB = 1 << 30;
  out.worker = { device: hello.device, version: hello.version };
  // The card, as the box will advertise it. vram_free comes from the driver
  // rather than from the worker's own ledger, so it already accounts for whatever
  // else the untrusted host is doing with the card -- on a desktop, an X server.
  out.card = {
    name: hello.device,
    vram_total_gb: hello.vram_total ? +(hello.vram_total / GB).toFixed(1) : 0,
    vram_free_gb: hello.vram_free != null ? +(hello.vram_free / GB).toFixed(1) : 0,
    vram_budget_gb: hello.vram_budget ? +(hello.vram_budget / GB).toFixed(1) : 0,
    sm_count: hello.sm_count || 0,
    capability: hello.capability || "",
    field_gmac_per_s: hello.field_gmac_per_s || 0,
  };

  const wqBytes = Buffer.from(wq.buffer, wq.byteOffset, wq.byteLength);
  const wdBytes = Buffer.from(wdHalf.buffer, wdHalf.byteOffset, wdHalf.byteLength);
  const wdOff = Math.ceil(wqBytes.length / 64) * 64;
  const wBufSize = wdOff + wdBytes.length;
  const xOff = 0, yOff = Math.ceil(3 * K / 64) * 64, aBufSize = yOff + N * 4;

  const wbid = Number(Buffer.from(await link.call(CMD.ALLOC_BUFFER, packAlloc(wBufSize, 'weights'))).readBigUInt64LE(0));
  const abid = Number(Buffer.from(await link.call(CMD.ALLOC_BUFFER, packAlloc(aBufSize, 'activations'))).readBigUInt64LE(0));
  await link.call(CMD.SET_TENSOR, packSet(wbid, 0, wqBytes));
  await link.call(CMD.SET_TENSOR, packSet(wbid, wdOff, wdBytes));

  // ---- the worker refuses a denylisted op, on the wire ---------------------
  {
    const probe = new ShieldedLink(HOST, PORT);
    await probe.connect();      // the worker is already up; no retry needed here
    await probe.call(CMD.HELLO, Buffer.from([1, 0, 0, 0]));
    await probe.call(CMD.ALLOC_BUFFER, packAlloc(4096, 'activations'));
    const bad = JSON.stringify({ nodes: [{ op: 'SOFT_MAX' }], outputs: [{ bid: 1, offset: 0, nbytes: 64 }] });
    try {
      await probe.call(CMD.GRAPH_INSTALL, Buffer.from(bad));
      out.denylist_refused = false;
    } catch (e) {
      out.denylist_refused = /SOFT_MAX|nonlinear/.test(e.message);
      out.denylist_reason = e.message.slice(0, 120);
    }
    probe.close();
  }

  const spec = {
    nodes: [{
      op: 'FIELD_GEMM', id: 'probe',
      wq: { bid: wbid, offset: 0 }, wd: { bid: wbid, offset: wdOff },
      x: { bid: abid, offset: xOff }, y: { bid: abid, offset: yOff },
      K, N, max_m: 1,
    }],
    outputs: [{ bid: abid, offset: yOff, nbytes: N * 4 }],
  };
  await link.call(CMD.GRAPH_INSTALL, Buffer.from(JSON.stringify(spec)));

  // ---- mask, offload, unmask, verify ---------------------------------------
  // Several rounds, each with a FRESH pad. Repetition is not padding out the
  // measurement: the first exchange against a new worker pays Triton's kernel
  // compilation for this shape, which is hundreds of milliseconds and has nothing
  // to do with either the network or the masking. Reporting one number would
  // either hide the compile or libel the transport, so both are reported.
  const ROUNDS = Number(arg('rounds', 5));
  const bank = new MaskBank();
  const times = [];
  let planes = null, sent = null, y = null, r = null, masked = null;

  for (let round = 0; round < ROUNDS; round++) {
    ({ r } = bank.issue(K));                 // one pad per plaintext, never reused
    masked = new Float64Array(K);
    for (let k = 0; k < K; k++) masked[k] = ((x[k] + r[k]) % M_MOD + M_MOD) % M_MOD;
    planes = toResidues(masked, K);

    sent = [];
    const frames = [];
    for (let p = 0; p < 3; p++) {
      const b = Buffer.from(planes[p].buffer, planes[p].byteOffset, planes[p].byteLength);
      sent.push(b);
      frames.push(frame(CMD.SET_TENSOR, packSet(abid, xOff + p * K, b)));
    }
    frames.push(frame(CMD.GRAPH_RECOMPUTE, packRecompute(0, 1)));
    frames.push(frame(CMD.GET_TENSOR, packRegion(abid, yOff, N * 4)));
    const t0 = process.hrtime.bigint();
    const resp = await link.exchange(frames);
    times.push(Number(process.hrtime.bigint() - t0) / 1e6);

    const yMaskedBuf = resp[resp.length - 1];
    const yMasked = new Float64Array(N);
    for (let j = 0; j < N; j++) yMasked[j] = yMaskedBuf.readInt32LE(j * 4);
    const u = refill(r, wFixed, K, N);
    y = new Float64Array(N);
    for (let j = 0; j < N; j++) { const v = ((yMasked[j] - u[j]) % M_MOD + M_MOD) % M_MOD; y[j] = v > HALF_M ? v - M_MOD : v; }
  }
  const warm = times.slice(1).sort((a, b) => a - b);
  out.rounds = ROUNDS;
  out.cold_round_trip_ms = +times[0].toFixed(3);          // includes kernel compile
  out.round_trip_ms = warm.length ? +warm[Math.floor(warm.length / 2)].toFixed(3)
                                  : out.cold_round_trip_ms;

  // reference: the plaintext product, computed here, never sent
  const want = new Float64Array(N);
  let peakY = 0;
  for (let j = 0; j < N; j++) {
    let acc = 0;
    for (let k = 0; k < K; k++) acc += x[k] * wFixed[k * N + j];
    want[j] = acc; peakY = Math.max(peakY, Math.abs(acc));
  }
  let exact = true;
  for (let j = 0; j < N; j++) if (y[j] !== want[j]) { exact = false; break; }
  out.exact = exact;
  out.peak_abs_y = peakY;
  out.field_headroom = +(HALF_M / peakY).toFixed(2);

  // The FIXTURE is deterministic so a failure reproduces from the log, but the
  // Freivalds secret is not part of the fixture: this probe's verdict is what
  // lets the box advertise shielded capacity, and "lie_rejected" is only a
  // security claim if the secret was unpredictable to the worker that answered.
  // So s comes from the class default (the OS CSPRNG), not from `rnd`.
  const fv = new Freivalds(wFixed, K, N);
  out.verified = fv.check(x, y);
  const lied = Float64Array.from(y); lied[N >> 1] += 1;      // single-element lie
  out.lie_rejected = !fv.check(x, lied);

  // ---- the adversary transcript, tested rather than asserted ---------------
  // An earlier version of this check counted residues that happened to EQUAL a
  // plaintext word and called the hits leakage. They are not: a residue lives in
  // [-125,125] and lands on any given small value about 1 time in 251, so a run
  // over 512 words x 3 planes is expected to collide once or twice by pure chance,
  // and it duly did. Coincidence is not correlation. What the pad actually
  // promises is that the transcript is independent of the secret, so that is what
  // is measured -- the same two tests reference/shielded_ref.py runs, against the
  // real bytes that crossed the socket.
  const transcript = Buffer.concat(sent);
  out.transcript_bytes = transcript.length;

  // (1) correlation between the masked field words and the plaintext.
  const meanOf = (a) => a.reduce((s, v) => s + v, 0) / a.length;
  const maskedArr = Array.from(masked), xArr = Array.from(x);
  const mx = meanOf(xArr), mm = meanOf(maskedArr);
  let num = 0, dx = 0, dm = 0;
  for (let k = 0; k < K; k++) {
    const a = xArr[k] - mx, b = maskedArr[k] - mm;
    num += a * b; dx += a * a; dm += b * b;
  }
  const corr = num / Math.sqrt(dx * dm);
  const nullBound = 3 / Math.sqrt(K);            // 3 sigma for a zero-correlation null
  out.transcript_correlation = +corr.toFixed(6);
  out.correlation_null_3sigma = +nullBound.toFixed(6);
  out.correlation_ok = Math.abs(corr) < nullBound;

  // (2) uniformity of the masked words over Z_M. A pad that is not uniform is not
  // a one-time pad, and a bank that silently reused one would show up here first.
  const BINS = 64;
  const hist = new Array(BINS).fill(0);
  for (let k = 0; k < K; k++) hist[Math.min(BINS - 1, Math.floor(masked[k] / M_MOD * BINS))]++;
  const expect = K / BINS;
  const chi2 = hist.reduce((s, h) => s + (h - expect) * (h - expect) / expect, 0);
  const CHI2_P001_DF63 = 103.4;                  // upper 0.1% point, 63 d.o.f.
  out.transcript_chi2 = +chi2.toFixed(1);
  out.chi2_threshold = CHI2_P001_DF63;
  out.uniform_ok = chi2 < CHI2_P001_DF63;
  out.bytes_out = link.bytesOut; out.bytes_in = link.bytesIn;

  link.close();
  out.ok = Boolean(out.exact && out.verified && out.lie_rejected &&
                   out.denylist_refused && out.correlation_ok && out.uniform_ok);
  console.log(JSON.stringify(out, null, 2));
  process.exit(out.ok ? 0 : 1);
}

main().catch((e) => { console.error('shielded probe FAILED:', e.message); process.exit(2); });
